core/src/main/scala/org/apache/spark/util/collection/PercentileHeap.scala

```diff
@@ -20,97 +20,55 @@ package org.apache.spark.util.collection
 import scala.collection.mutable.PriorityQueue
 
 /**
- * PercentileHeap is designed to be used to quickly track the percentile of a group of numbers
- * that may contain duplicates. Inserting a new number has O(log n) time complexity and
- * determining the percentile has O(1) time complexity.
- * The basic idea is to maintain two heaps: a smallerHalf and a largerHalf. The smallerHalf
- * stores the smaller half of all numbers while the largerHalf stores the larger half.
- * The sizes of two heaps need to match the percentage each time when a new number is inserted so
- * that the ratio of their sizes is percentage to (1 - percentage). Therefore each time when
- * percentile() is called we check if the sizes of two heaps match the percentage. If they do,
- * we should return the average of the two top values of heaps. Otherwise we return the top of the
- * heap which exceeds its percentage.
+ * PercentileHeap tracks the percentile of a collection of numbers.
+ *
+ * Insertion is O(log n), Lookup is O(1).
+ *
+ * The implementation keeps two heaps: a bottom heap (`botHeap`) and a top heap (`topHeap`). The
+ * bottom heap stores all the numbers below the percentile and the top heap the ones above the
+ * percentile. During insertion the relative sizes of the heaps are adjusted to match the
+ * target percentile.
  */
-private[spark] class PercentileHeap(percentage: Double = 0.5)(implicit val ord: Ordering[Double]) {
-  assert(percentage >= 0 && percentage <= 1)
+private[spark] class PercentileHeap(percentage: Double = 0.5) {
+  assert(percentage > 0 && percentage < 1)
 
-  /**
-   * Stores all the numbers less than the current percentile in a smallerHalf,
-   * i.e percentile is the maximum, at the root.
-   */
-  private[this] val smallerHalf = PriorityQueue.empty[Double](ord)
+  private[this] val topHeap = PriorityQueue.empty[Double](Ordering[Double].reverse)
+  private[this] val botHeap = PriorityQueue.empty[Double](Ordering[Double])
 
-  /**
-   * Stores all the numbers greater than the current percentile in a largerHalf,
-   * i.e percentile is the minimum, at the root.
-   */
-  private[this] val largerHalf = PriorityQueue.empty[Double](ord.reverse)
+  def isEmpty(): Boolean = botHeap.isEmpty && topHeap.isEmpty
 
-  def isEmpty(): Boolean = {
-    smallerHalf.isEmpty && largerHalf.isEmpty
-  }
+  def size(): Int = botHeap.size + topHeap.size
 
-  def size(): Int = {
-    smallerHalf.size + largerHalf.size
+  /**
+   * Returns percentile of the inserted elements as if the inserted elements were sorted and we
+   * returned `sorted((sorted.length * percentage).toInt)`.
+   */
+  def percentile(): Double = {
+    if (isEmpty) throw new IndexOutOfBoundsException("empty")
```

Comment:
I tested it locally, `PriorityQueue` throws `NoSuchElementException` if dequeue with empty data. We should keep using `NoSuchElementException` here.

Reply:
Done.